### PR TITLE
Improve templating

### PR DESCRIPTION
**Pull request by greg0ire**
While working on #9528, which was getting bigger and bigger, I noticed some improvements that I feel deserve a separate PR, given the complexity.

### CODE REVIEW

File: lib/Doctrine/ORM/Mapping/Driver/AttributeReader.php [hunk 1]
```diff
@@ -14,58 +15,93 @@
 use function assert;
 use function is_string;
 use function is_subclass_of;
+use function sprintf;
 
 /**
  * @internal
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Mapping/Driver/AttributeReader.php:21]: This allows me to do all sorts of BC-breaks.

**Comment by greg0ire** [comment]
The drop in coverage looks too high to be realistic. Let me write some tests and try again, to see how it affects the coverage.

**Comment by greg0ire** [comment]
> compared to 5a55772 

Maybe that's the issue… this is a 1-week old ancestor of 2.12.x

EDIT: looks like my commit was based on that commit, and I just needed to rebase lol

**Comment by greg0ire** [comment]
The codecov comments seem to relate to a bad handling of multiline expressions. When I try using `PCOV` locally, I get this gap:

```xml
      <line nr="68">
        <covered by="Doctrine\Tests\ORM\Mapping\AttributeReaderTest::testItThrowsWhenGettingRepeatableAnnotationWithTheWrongMethod"/>
      </line>
      <line nr="74">

```

File: lib/Doctrine/ORM/Mapping/Driver/AttributeReader.php [hunk 2]
```diff
@@ -14,58 +15,93 @@
...
  */
 final class AttributeReader
 {
-    /** @var array<string,bool> */
+    /** @var array<class-string<Annotation>,bool> */
     private array $isRepeatableAttribute = [];
 
-    /** @return array<Annotation|RepeatableAttributeCollection> */
+    /**
+     * @return class-string-map<T, T|RepeatableAttributeCollection<T>>
```
Comment (derrabus) [inline lib/Doctrine/ORM/Mapping/Driver/AttributeReader.php]: I didn't know `class-string-map` – which tool supports that type?

Also, I think we usually use a vendor-prefixed annotation like `@psalm-return` for types that not every tool might understand.

File: lib/Doctrine/ORM/Mapping/Driver/AttributeReader.php [hunk 3]
```diff
@@ -14,58 +15,93 @@
...
+     *
+     * @return T|null
+     *
+     * @template T of Annotation
+     */
     public function getPropertyAnnotation(ReflectionProperty $property, $annotationName)
     {
+        if ($this->isRepeatable($annotationName)) {
+            throw new LogicException(sprintf(
+                'The annotation "%s" is repeatable. Call getPropertyAnnotationCollection() instead.',
```
Comment (derrabus) [inline lib/Doctrine/ORM/Mapping/Driver/AttributeReader.php]:
```suggestion
                'The attribute "%s" is repeatable. Call getPropertyAnnotationCollection() instead.',
```

File: tests/Doctrine/Tests/ORM/Mapping/AttributeReaderTest.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+use Doctrine\ORM\Mapping as ORM;
+use Doctrine\ORM\Mapping\Driver\AttributeReader;
+use LogicException;
+use PHPUnit\Framework\TestCase;
+use ReflectionProperty;
+
+class AttributeReaderTest extends TestCase
+{
+    /**
+     * @requires PHP 8.0
```
Comment (derrabus) [inline tests/Doctrine/Tests/ORM/Mapping/AttributeReaderTest.php]: I think we can move this annotation to the class level.

File: lib/Doctrine/ORM/Mapping/Driver/AttributeReader.php (hunk 2 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/Driver/AttributeReader.php]: This is supported by Psalm: https://github.com/vimeo/psalm/search?q=class-string-map
It does not seem supported by PHPStan (I could not find an occurrence in the codebase).

File: tests/Doctrine/Tests/ORM/Mapping/AttributeReaderTest.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        $this->expectException(LogicException::class);
+        $this->expectExceptionMessage(
+            'The attribute "Doctrine\ORM\Mapping\Index" is repeatable. Call getPropertyAnnotationCollection() instead.'
+        );
+        $reader->getPropertyAnnotation($property, ORM\Index::class);
+    }
+
+    /**
+     * @requires PHP 8.0
+     */
```
Comment (derrabus) [inline tests/Doctrine/Tests/ORM/Mapping/AttributeReaderTest.php]:
```suggestion
```

**Comment by derrabus** [comment]
LGTM

Please merge, once the redundant doc block that I commented on is removed. 🙂

**Review by derrabus — APPROVED** [review]

**Pull request merged**